### ISSUE TITLE
Leverage TensorDict batch_size semantics with torch.vmap

### ISSUE THREAD

**Issue by vmoens** [vmoens/mujoco-torch#23]
## Motivation

Currently, `MjTensorClass` (the `TensorClass` base for `Model`, `Data`, `Contact`, etc.) is used as a **pytree-compatible named struct** — for `replace()`, `.to(device)`, `torch.stack()`, `tree_map` — but the TensorDict `batch_size` semantics are essentially unused for `Model`/`Data`. It is only used structurally for `Contact` slicing.

- `Model` and `Data` are always created with `batch_size=[]` at `device_put` time.
- When the user `torch.stack`s multiple `Data` objects for vmap, the stacking adds a leading dimension to every tensor, but the `batch_size` metadata isn't updated to reflect that.
- `vmap` operates on the raw tensor dimensions via pytree flattening, not on TensorDict's batch semantics.
- `Contact` and `_Efc` use `batch_size=[ncon]` / `[nefc]` to track row counts, but after vmap the batch_size can get out of sync, requiring `auto_batch_size_()` fixups.

TensorDict supports being treated as a single "tensor" under `torch.vmap` — meaning you can vmap over a TensorDict's batch dimension without decomposing it into individual tensors. The batching is automatic and the TensorDict is not flattened into its constituent leaves. This is cheaper at trace time and produces a cleaner graph.

There are several concrete opportunities to leverage this.

---

## 1. User-level batching: `Data` with `batch_size=[B]`

Today the user-facing pattern is:

```python
envs = [mujoco_torch.device_put(mujoco.MjData(m_mj)) for _ in range(B)]
d_batch = torch.stack(envs)                          # batch_size stays []
vmap_step = torch.vmap(lambda d: step(mx, d))        # vmap over dim 0 of every tensor
```

`torch.stack` on TensorClass does add a leading dimension to every tensor, but `batch_size` stays `[]` — the TensorDict doesn't *know* it's batched. If it were constructed with `batch_size=[B]`, you could:

- **Skip the vmap decomposition entirely.** TensorDict registers itself with functorch so that `torch.vmap` over a TensorDict with `batch_size=[B]` treats the whole thing as a single object whose leading dim is the batch — no pytree flattening into individual tensors, no per-leaf `in_dims` tracking. This is cheaper at trace time and produces a cleaner graph.
- **Use `d_batch[i]` for indexing** with proper batch semantics, `d_batch.apply(fn)` for element-wise ops, and `d_batch.reshape(B1, B2)` for multi-level batching (e.g. `num_envs × num_agents`).

The change would be in `device_put` (or a helper) and in how `torch.stack` is called — making sure the result has `batch_size=[B]`.

---

## 2. Internal vmap over `Contact` — eliminate `auto_batch_size_()` fixups

The constraint code already uses `batch_size` on `Contact`:

```python
# constraint.py
if contact.batch_size != torch.Size([contact.dist.shape[0]]):
    contact = contact.clone(recurse=False)
    contact.auto_batch_size_()
contact = contact[:ncon_fl]
```

Then it does `@torch.vmap` over the sliced `Contact`. But the `auto_batch_size_()` fixup suggests the `batch_size` is getting lost somewhere upstream — likely because `collision_driver.collision` builds `Contact` with `batch_size=[ncon]` but after vmap over environments the extra dim doesn't propagate correctly.

If `Contact` always had a correct `batch_size`, we could drop the `auto_batch_size_()` fixups entirely.

---

## 3. `_Efc` — constraint rows as batched TensorClass

In `constraint.py`, every `_instantiate_*` function manually constructs an `_Efc` with `batch_size=[j.shape[0]]`, then `@torch.vmap` iterates over it. Since `_Efc` already declares its `batch_size`, the vmap could leverage that directly — the `@torch.vmap` decorator on a function taking `_Efc` would automatically batch/unbatch along the `_Efc`'s batch dimension rather than decomposing it into individual tensors.

---

## 4. `GeomInfo` / `ConvexInfo` in collision — eliminate manual `in_axes`

In `collision_driver.py`:

```python
g1, g2, in_axes = _pair_info(m, d, ...)
res = torch.vmap(fn, in_axes)(g1, g2)
```

Here `g1` and `g2` are `GeomInfo` TensorClasses, and `in_axes` is manually constructed to tell vmap which dims to batch over. If `GeomInfo` had a proper `batch_size=[n_pairs]`, we wouldn't need the manual `in_axes` — vmap would know the batch dim from the TensorClass itself.

---

## 5. `scan.flat` and `scan.body_tree` — simplify `_nvmap`

`_nvmap` exists specifically because `torch.vmap` can't take `None` or numpy args. But TensorDict's vmap integration handles the pytree registration differently — if the grouped data were packed into a TensorClass with `batch_size=[group_size]`, you could vmap over the TensorClass directly, and numpy static args would live outside the TensorClass (not as vmap inputs at all). This would let us simplify or drop `_nvmap` in favour of plain `torch.vmap`.

---

## 6. `xfrc_accumulate` — body-level vmap

Currently:

```python
# support.py
qfrc = torch.vmap(apply_ft, (None, None, 0, 0, 0, 0))(
    m, d, d.xfrc_applied[:, :3], d.xfrc_applied[:, 3:], d.xipos, body_ids)
```

The `(None, None, 0, 0, 0, 0)` `in_dims` is fragile — it manually says "don't batch `m` and `d`, batch the rest along dim 0". If the force/torque/pos data were packed into a TensorClass with `batch_size=[nbody]`, we'd just vmap over that single object.

---

## Priority

The biggest wins would be:

1. **User-level `Data` batching** — most visible API surface. Making `torch.stack(envs)` produce a `Data` with `batch_size=[B]` simplifies the vmap story (no lambda needed, no pytree flattening of potentially dozens of fields), produces cleaner compile graphs, and enables `d_batch[i]` / `d_batch.reshape(...)` / `d_batch.unbind()` for free.

2. **`Contact` batch consistency** — eliminating the `auto_batch_size_()` fixups by ensuring `Contact.batch_size` stays correct through the entire collision → constraint pipeline.

3. **Collision `in_axes` elimination** — letting `GeomInfo.batch_size` drive the vmap dim instead of manually computing `in_axes`.

The deeper internal uses (scan, `_Efc`) are more incremental since they're already working, but they'd become cleaner and less error-prone with proper `batch_size` semantics.
